engine/tests/phpunit/Elgg/TestCase.php

```diff
@@ -0,0 +1,165 @@
+<?php
+
+namespace Elgg;
+
+use Elgg\Di\ServiceProvider;
+use Elgg\Http\Request;
+use ElggDiskFilestore;
+use stdClass;
+use Zend\Mail\Transport\InMemory as InMemoryTransport;
+
+abstract class TestCase extends \PHPUnit_Framework_TestCase {
+
+	/**
+	 * @var Application
+	 */
+	static $_app;
+
+	/**
+	 * @var Config
+	 */
+	static $_config;
+
+	/**
+	 * Constructor
+	 *
+	 * @global stdClass $CONFIG Global config
+	 * @global stdClass $_ELGG  Global vars
+	 */
+	public function __construct($name = NULL, array $data = array(), $dataName = '') {
+
+		date_default_timezone_set('America/Los_Angeles');
+
+		error_reporting(E_ALL | E_STRICT);
+
+		/**
+		 * This is here as a temporary solution only. Instead of adding more global
+		 * state to this file as we migrate tests, try to refactor the code to be
+		 * testable without global state.
+		 */
+		global $CONFIG;
+		$CONFIG = (object) [
+					'Config_file' => false,
+					'dbprefix' => 'elgg_',
+					'boot_complete' => false,
+					'wwwroot' => 'http://localhost/',
+					'path' => __DIR__ . '/../../../../',
+					'dataroot' => __DIR__ . '/../test_files/dataroot/',
+					'cacheroot' => __DIR__ . '/../test_files/cacheroot/',
+					'site_guid' => 1,
+					'AutoloaderManager_skip_storage' => true,
+					'simplecache_enabled' => false,
+					'system_cache_enabled' => false,
+					'Elgg\Application_phpunit' => true,
+					// \Elgg\Config::get() falls back to loading config values from database
+					// for undefined keys. This flag ensures we do not attempt reading data
+					// from database during tests
+					'site_config_loaded' => true,
+					'icon_sizes' => array(
+						'topbar' => array('w' => 16, 'h' => 16, 'square' => true, 'upscale' => true),
+						'tiny' => array('w' => 25, 'h' => 25, 'square' => true, 'upscale' => true),
+						'small' => array('w' => 40, 'h' => 40, 'square' => true, 'upscale' => true),
+						'medium' => array('w' => 100, 'h' => 100, 'square' => true, 'upscale' => true),
+						'large' => array('w' => 200, 'h' => 200, 'square' => false, 'upscale' => false),
+						'master' => array('w' => 550, 'h' => 550, 'square' => false, 'upscale' => false),
+					),
+		];
+
+		global $_ELGG;
+		$_ELGG = (object) [
+					'view_path' => __DIR__ . '/../../../../views/',
+					'allowed_ajax_views' => [],
+		];
+
+
+		$config = new Config($CONFIG);
+		$this->config($config);
+
+		$sp = new ServiceProvider($config);
+
+		$sp->setValue('mailer', new InMemoryTransport());
+
+		$sp->siteSecret->setTestingSecret('z1234567890123456789012345678901');
+
+		$app = new Application($sp);
+		Application::$_instance = $app;
+		$app->loadCore();
+
```

Comment:
With this we *should* have fewer leftover state problems, not surely not a complete fix.